File: #0725.split-linked-list-in-parts.js
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        
        ListNode* cur = head;
        int n = 0;
        while (cur) {
            n++;
            cur = cur->next;
        }
        
        int width = n / k;
        int rem = n % k;
        vector<ListNode*> result(k, NULL);
        cur = head;
        for (int i = 0; i < k; i++) {
            ListNode* head = cur;
            ListNode* prev = cur;
            for (int j = 0; j < width + (i < rem ? 1 : 0) - 1; j++) {
                if (prev) {
                    prev = prev->next;
                }
            }

            if (prev) {
                cur = prev->next;
                prev->next = NULL;
            }
            result[i] = head;
        }
        return result;
    }
};

File: #0725.split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {

        int length = 0;
        ListNode temp = head;
        while (temp != null) {
            length++;
            temp = temp.next;
        }
        temp = head;

        int partLength = length / k;
        int extraNodes = length % k;
        ListNode[] arr = new ListNode[k];
        ListNode prev = null;
        for (int i = 0; i < arr.length && temp != null; i++, extraNodes--) {
            arr[i] = temp;
            for (int j = 0; j < partLength + (extraNodes > 0 ? 1 : 0)
                && temp != null; j++) {
                prev = temp;
                temp = temp.next;
            }
            prev.next = null;
        }
        return arr;
    }
}
